MapboxNavigation/InstructionPresenter.swift

```diff
@@ -14,26 +14,58 @@ class InstructionPresenter {
-        }
-
         let string = NSMutableAttributedString()
-
-        for component in instruction {
+        fittedAttributedComponents().forEach { string.append($0) }
+        return string
+    }
+    
+    func fittedAttributedComponents() -> [NSAttributedString] {
+        guard let label = self.label else { return [] }
+        var attributedComponents = self.attributedComponents()
+        let availableBounds = label.availableBounds()
+        let totalWidth = attributedComponents.map { $0.size() }.reduce(.zero, +).width
+        let stringFits = totalWidth <= availableBounds.width
+        
+        guard !stringFits else { return attributedComponents }
+        
+        let indexedComponents = instruction.enumerated().map { IndexedVisualInstructionComponent(component: $1, index: $0) }
+        let filtered = indexedComponents.filter { $0.component.abbreviation != nil }
+        let sorted = filtered.sorted { $0.component.abbreviationPriority < $1.component.abbreviationPriority }
+        for component in sorted {
+            let isFirst = component.index == 0
+            let joinChar = isFirst ? "" : " "
+            guard component.component.type == .text else { continue }
+            guard let abbreviation = component.component.abbreviation else { continue }
+            
+            attributedComponents[component.index] = NSAttributedString(string: joinChar + abbreviation, attributes: attributesForLabel(label))
+            let newWidth = attributedComponents.map { $0.size() }.reduce(.zero, +).width
+            
+            if newWidth <= availableBounds.width {
+                break
+            }
+        }
+        
+        return attributedComponents
+    }
+    
+    func attributedComponents() -> [NSAttributedString] {
+        guard let label = self.label else { return [NSAttributedString()] }
+        var strings = [NSAttributedString]()
+        let components = instruction
+        
+        for component in components {
             let isFirst = component == instruction.first
-            let joinChar = !isFirst ? " " : ""
-
+            let joinChar = isFirst ? "" : " "
+            
             if let shieldKey = component.shieldKey() {
                 if let cachedImage = imageRepository.cachedImageForKey(shieldKey) {
-                    string.append(NSAttributedString(string: joinChar))
-                    string.append(attributedString(withFont: label.font, shieldImage: cachedImage))
+                    strings.append(attributedString(withFont: label.font, shieldImage: cachedImage))
                 } else {
                     // Display road code while shield is downloaded
                     if let text = component.text {
-                        string.append(NSAttributedString(string: joinChar + text, attributes: attributesForLabel(label)))
+                        strings.append(NSAttributedString(string: joinChar + text, attributes: attributesForLabel(label)))
```

Comment:
`strings.append(NSAttributedString(string: "\(joinChar)\(text)", attributes: attributesForLabel(label)))`

Reply:
Reason: As we continuously concatenate the strings in a for-loop, we creating multiple strings to get to the final string. I see we also map the results as well. Just an observation as I reviewed the scope of the function `attributedComponents()` and its callee(s).

Reply:
I fail to see the distinction. It's just another syntax for concatenating using the string concat operator(+) or string formatting but the reason indicates that it's more than just the code style?

We do one full iteration first to stringify all components including shields.
"[I-280] West Fremont Avenue"
then one more iteration until the instruction fits on the screen
"[I-280] W Fremont Avenue"
if West had the highest priority.

Reply:
@frederoni difference is negligible with simple code. However, as we iterate, a larger data set it may affect performance. Interpolation does make good use of memory as stated in the abstract. Apple offers this as a given, so why not use it 😄 -- just like how we prefer to use functional programming 🤷‍♂️ 

cc @bsudekum

Reply:
A quick performance test indicated that using the + operator for concatenating strings is 19 × faster than concatenating using string interpolation.

Reply:
@frederoni I believe this was more about good use of memory. However, I see how my last mention of performance could have stirred us towards performance measurement.